python/adelphi/adelphi/export.py

```diff
@@ -81,9 +82,23 @@ def get_keyspaces(self, cluster, props):
 
         log.info("Processing the following keyspaces: %s", ','.join((ks.name for ks in keyspaces)))
 
-        # anonymize_keyspace mutates keyspace state so we must trap keyspace_id before we (possibly) call it
-        base_map = { ks : self.build_keyspace_id(ks) for ks in keyspaces}
-        return base_map if not props['anonymize'] else {anonymize_keyspace(ks) : keyspace_id for (ks, keyspace_id) in base_map.items()}
+        def process_keyspace(ks):
+            # Anonymize keyspace first.  This has to be done before any re-ordering since
+            # names may/will change as a result of this process
+            if props['anonymize']:
+                anonymize_keyspace(ks)
+
+            # Make sure tables, UDTs and indexes are ordered by name
+            tables = ks.tables.items()
+            ks.tables = OrderedDict(sorted(tables, key=lambda x: x[0]))
+            user_types = ks.user_types.items()
+            ks.user_types = OrderedDict(sorted(user_types, key=lambda x:x[0]))
+            indexes = ks.indexes.items()
+            ks.indexes = OrderedDict(sorted(indexes, key=lambda x:x[0]))
+
+            return ks
+
+        return {process_keyspace(ks) : self.build_keyspace_id(ks) for ks in keyspaces}
```

Comment:
While testing I started seeing diff failures based on ordering of tables within the keyspace.  The root cause here appeared to be a reliance on platform-specific functionality somewhere in the code combined with subtle behavioural differences on 2.7 vs. Python 3 in some weird cases.  I never did identify exactly what the platform-specific code was or why it was triggered in some cases but not others... in fact I could never reproduce it in practice.  But it consistently caused my tests to fail.

This implementation explicitly addresses the problem by enforcing an ordering based on name for tables, UDTs and indexes.  We do it after anonymization since that process might change those names.

Reply:
Update: enforcement of explicit ordering within the *Metadata objects has been abandoned in favor of a new comparison strategy based on sha256 hashes for each CQL command in a given schema.